src/Microsoft.VisualStudio.Threading/ProgressWithCompletion.cs

```diff
@@ -81,37 +139,85 @@ void IProgress<T>.Report(T value)
         /// <param name="value">The value representing the updated progress.</param>
         protected virtual void Report(T value)
         {
-#pragma warning disable CA2008 // Do not create tasks without passing a TaskScheduler
-            var reported = this.taskFactory.StartNew(() => this.handler(value)).Unwrap();
-#pragma warning restore CA2008 // Do not create tasks without passing a TaskScheduler
-            lock (this.syncObject)
+            if (this.joinableTaskFactory != null)
             {
-                this.outstandingTasks.Add(reported);
-            }
+                JoinableTask joinableTask = this.joinableTaskFactory.RunAsync(
+                    async delegate
+                    {
+                        // Emulate the behavior of having captured a SynchronizationContext by invoking the handler on the main thread
+                        // if the constructor was on the main thread. Otherwise use the threadpool thread. But never invoke the handler
+                        // inline with our caller, per the behavior folks expect from this and .NET's Progress<T> class.
+                        if (this.createdOnMainThread)
+                        {
+                            await this.joinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true);
+                        }
+                        else
+                        {
+                            await TaskScheduler.Default.SwitchTo(alwaysYield: true);
+                        }
 
-            reported.ContinueWith(
-                t =>
+                        await this.handler(value).ConfigureAwaitRunInline();
+                    });
+                this.outstandingJoinableTasks.Add(joinableTask);
```

Comment:
Does this list get cleaned up internally for completed joinable tasks?

Reply:
Yes. `JoinableTaskCollection` clears itself automatically.